File: nand2tetris/08/codeWriter/codeWriter.go
/**
  @Author: HuangChuan
  @Date: 2021/12/27 15:32
  @Description:// Translate the VM command to assemble code
**/
package codeWriter

import (
	"bufio"
	"fmt"
	"io"
	"nand2tetris/08/parser"
)

type CodeWriter struct {
	writer   io.Writer
	addr     int
	callcnt  int
	filename string
}

//
//  NewCodeWriter
//  @Description: Constructor
//  @param w
//  @return *CodeWriter
//
func NewCodeWriter(w io.Writer) *CodeWriter {
	cw := &CodeWriter{
		writer: w,
	}
	cw.writeInit()
	return cw
}

//
//  SetFileName
//  @Description: set the filename of output file
//  @receiver cw
//  @param filename
//
func (cw *CodeWriter) SetFileName(filename string) {
	cw.filename = filename
}

//
//  WriteArithmetic
//  @Description: write assemble code of arithmetic commands
//  @receiver cw
//  @param cmd
//
func (cw *CodeWriter) WriteArithmetic(cmd string) {
	switch cmd {
	case "add":
		cw.writeAddSubAndOr("M=D+M")
	case "sub":
		cw.writeAddSubAndOr("M=M-D")
	case "and":
		cw.writeAddSubAndOr("M=D&M")
	case "or":
		cw.writeAddSubAndOr("M=D|M")
	case "neg":
		cw.writeNegNot("M=-M")
	case "not":
		cw.writeNegNot("M=!M")
	case "eq":
		cw.writeEqGtLt("JEQ")
	case "gt":
		cw.writeEqGtLt("JGT")
	case "lt":
		cw.writeEqGtLt("JLT")
	}
}

//
//  WritePushPop
//  @Description: write assemble code of push and pop commands
//  @receiver cw
//  @param cmd
//  @param segment
//  @param index
//
func (cw *CodeWriter) WritePushPop(cmd parser.CommandType, segment string, index int) {
	switch cmd {
	case parser.C_PUSH:
		switch segment {
		case "argument":
			cw.writePushSymbol("ARG", index)
		case "local":
			cw.writePushSymbol("LCL", index)
		case "this":
			cw.writePushSymbol("THIS", index)
		case "that":
			cw.writePushSymbol("THAT", index)
		case "constant":
			cw.writePushConstant(index)
		case "pointer":
			cw.writePushRegister(index + 3)
		case "temp":
			cw.writePushRegister(index + 5)
		case "static":
			cw.writePushStatic(index)
		}
	case parser.C_POP:
		switch segment {
		case "argument":
			cw.writePopSymbol("ARG", index)
		case "local":
			cw.writePopSymbol("LCL", index)
		case "this":
			cw.writePopSymbol("THIS", index)
		case "that":
			cw.writePopSymbol("THAT", index)
		case "pointer":
			cw.writePopRegister(index + 3)
		case "temp":
			cw.writePopRegister(index + 5)
		case "static":
			cw.writePopStatic(index)
		}
	}

}

//
//  WriteLabel
//  @Description: write assemble code of label command
//  @receiver cw
//  @param label
//
func (cw *CodeWriter) WriteLabel(label string) {
	symbol := fmt.Sprintf("%s$%s", cw.filename, label)
	asm := `
(%s)
`
	asm = fmt.Sprintf(asm, symbol)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
}

//
//  WriteIf
//  @Description: write assemble code of if-goto
//  @receiver cw
//  @param label
//
func (cw *CodeWriter) WriteIf(label string) {
	symbol := fmt.Sprintf("%s$%s", cw.filename, label)
	asm := `
@SP
M=M-1
@SP
A=M
D=M
@%s
D;JNE
`
	asm = fmt.Sprintf(asm, symbol)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
}

//
//  WriteGoto
//  @Description: write assemble code of goto command
//  @receiver cw
//  @param label
//
func (cw *CodeWriter) WriteGoto(label string) {
	symbol := fmt.Sprintf("%s$%s", cw.filename, label)
	asm := `
@%s
0;JMP
`
	asm = fmt.Sprintf(asm, symbol)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
}

//
//  WriteReturn
//  @Description: write assemble code of return command
//  @receiver cw
//
func (cw *CodeWriter) WriteReturn() {
	asm := `
@LCL
D=M
@R13
M=D
@5
D=A
@R13
A=M-D
D=M
@R14
M=D
@SP
M=M-1
@SP
A=M
D=M
@ARG
A=M
M=D
@ARG
D=M+1
@SP
M=D
@1
D=A
@R13
A=M-D
D=M
@THAT
M=D
@2
D=A
@R13
A=M-D
D=M
@THIS
M=D
@3
D=A
@R13
A=M-D
D=M
@ARG
M=D
@4
D=A
@R13
A=M-D
D=M
@LCL
M=D
@R14
A=M
0;JMP
`
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
}

//
//  WriteFunction
//  @Description: write assemble code of function
//  @receiver cw
//  @param functionName
//  @param numArgs
//
func (cw *CodeWriter) WriteFunction(functionName string, numArgs int) {
	asm := `
(%s)
`
	asm = fmt.Sprintf(asm, functionName)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
	// init local variable to 0
	for i := 0; i < numArgs; i++ {
		cw.writePushConstant(0)
	}
}

//
//  WriteCall
//  @Description: write assemble code of call command
//  @receiver cw
//  @param functionName
//  @param numArgs
//
func (cw *CodeWriter) WriteCall(functionName string, numArgs int) {
	rAddr := fmt.Sprintf("%s$%d", functionName, cw.callcnt)
	asm := `
@%s
D=A
@SP
A=M
M=D
@SP
M=M+1
`
	asm = fmt.Sprintf(asm, rAddr)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
	// push LCL, ARG, THIS, THAT
	cw.writePushRegisterByName("LCL")
	cw.writePushRegisterByName("ARG")
	cw.writePushRegisterByName("THIS")
	cw.writePushRegisterByName("THAT")
	asm2 := `
@SP
D=M
@%d
D=D-A
@5
D=D-A
@ARG
M=D
@SP
D=M
@LCL
M=D
@%s
0;JMP
(%s)
`
	asm2 = fmt.Sprintf(asm2, numArgs, functionName, rAddr)
	w = bufio.NewWriter(cw.writer)
	w.WriteString(asm2)
	w.Flush()
	cw.callcnt++
}

//
//  writeInit
//  @Description: init code, init the stack pointer
//  @receiver cw
//
func (cw *CodeWriter) writeInit() {
	asm := `
@256
D=A
@SP
M=D
`
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
	cw.WriteCall("Sys.init", 0)
}

//
//  writeAddSub
//  @Description: write assemble code of add, sub, and, or command
//  @receiver cw
//  @param op
//
func (cw *CodeWriter) writeAddSubAndOr(op string) {
	/*
		1. pop to M register, decrease stack pointer by one.
			- @SP
			- M=M-1
		2. push to D register from M register, get the second num.
			- A=M
			- D=M
		3. pop to M register, decrease stack pointer by one.
			- @SP
			- M=M-1
		4. Add M register and D register, push to M register, get the first num
			- A=M
			- M=D+M
		5. increase stack pointer by one.（Initialize stack pointer）
			- @SP
			- M=M+1
	*/
	asm := `
@SP
M=M-1
A=M
D=M
@SP
M=M-1
A=M
%s
@SP
M=M+1
`
	asm = fmt.Sprintf(asm, op)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
}

//
//  writeNegNot
//  @Description: write assemble code of neg, not command.
//  @receiver cw
//  @param op
//
func (cw *CodeWriter) writeNegNot(op string) {
	/*
		1. pop to M register, decrease stack pointer by one.
			- @SP
			- M=M-1
		2. Make it a negative number at M register.
			- A=M
			- M=-M
		3. increase stack pointer by one.（Initialize stack pointer）
			- @SP
			- M=M+1
	*/
	asm := `
@SP
M=M-1
A=M
%s
@SP
M=M+1
`
	asm = fmt.Sprintf(asm, op)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
}

//
//  writeEqGtLt
//  @Description: write assemble code of eq, gt, lt command
//  @receiver cw
//  @param op
//
func (cw *CodeWriter) writeEqGtLt(op string) {
	/*
		1. pop to M register, decrease stack pointer by one.
			- @SP
			- M=M-1
		2. push to D register from M register.
			- A=M
			- D=M
		3. pop to M register, decrease stack pointer by one.
			- @SP
			- M=M-1
		4. subtract M register from D register, push to D register.
			- A=M
			- D=M-D
		5. Set TRUE to the memory pointed to by the stack pointer
			- @SP
			- A=M
			- M=-1 //0xFFFF
		6. if D register is 0, jump to LABEL1
			- @LABEL1
			- D;JEQ
		7. Set FALSE to the memory pointed to by the stack pointer
			- @SP
			- A=M
			- M=0 //0x0000
		8. Jump destination label
			- (LABEL1)
		9. increase stack pointer by one.（Initialize stack pointer）
			- @SP
			- M=M+1
	*/
	asm := `
@SP
M=M-1
A=M
D=M
@SP
M=M-1
A=M
D=M-D
@SP
A=M
M=-1
@LABEL%d
D;%s
@SP
A=M
M=0
(LABEL%d)
@SP
M=M+1
`
	asm = fmt.Sprintf(asm, cw.addr, op, cw.addr)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
}

//
//  writePushSymbol
//  @Description: write assemble code of push symbol command
//  @receiver cw
//  @param symbol
//  @param index
//
func (cw *CodeWriter) writePushSymbol(symbol string, index int) {
	/*
		1. put 0 in D register
			- @0
			- D=A
		2. Add the stack area pointed to by LCL(local) and the D register
			- @LCL
			- D=D+M
		3. Save D register in the stack area pointed to by R13(temp)
			- @R13
			- M=D
		4. Set the stack area pointed to by R13 in the A register
			- A=M
			- D=M
		5. Save D register in the stack area pointed to by stack pointer
			- @SP
			- A=M
			- M=D
		6. increase stack pointer by one.（Initialize stack pointer）
			- @SP
			- M=M+1
	*/
	asm := `
@%d
D=A
@%s
D=D+M
@R13
M=D
A=M
D=M
@SP
A=M
M=D
@SP
M=M+1
`
	asm = fmt.Sprintf(asm, index, symbol)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
}

//
//  writePushRegister
//  @Description: write the assemble code of push register value into stack pointer
//  @receiver cw
//  @param r
//
func (cw *CodeWriter) writePushRegister(r int) {
	asm := `
@R%d
D=M
@SP
A=M
M=D
@SP
M=M+1
`
	asm = fmt.Sprintf(asm, r)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()

}

//
//  writePushRegisterByName
//  @Description: write assemble code of push register value into stack pointer
//  @receiver cw
//  @param register
//
func (cw *CodeWriter) writePushRegisterByName(register string) {
	asm := `
@%s
D=M
@SP
A=M
M=D
@SP
M=M+1
`
	asm = fmt.Sprintf(asm, register)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
}

//
//  writePushConstant
//  @Description: write assemble code of push constant value into stack pointer
//  @receiver cw
//  @param index
//
func (cw *CodeWriter) writePushConstant(index int) {
	/*
		1. put 0 in D register
			- @0
			- D=A
		2. pop to M register from D register. M register is top +1 element in stack
			- @SP
			- M=A //Set empty value, using stack pointer
			- M=D
		3. increase stack pointer by one.（Initialize stack pointer）
			- @SP
			- M=M+1
	*/
	asm := `
@%d
D=A
@SP
A=M
M=D
@SP
M=M+1
`
	asm = fmt.Sprintf(asm, index)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
}

//
//  writePushStatic
//  @Description: write assemble code of push static value into stack pointer
//  @receiver cw
//  @param index
//
func (cw *CodeWriter) writePushStatic(index int) {
	/*
		1. Put Static value in D register.
			- @StaticTest.1
			- D=M
		2. Add D register in the stack area pointed to by stack pointer
			- @SP
			- A=M
			- M=D
		3. increase stack pointer by one.（Initialize stack pointer）
			- @SP
			- M=M+1
	*/
	// set the name of static value
	static := fmt.Sprintf("%s.%d", cw.filename, index)
	asm := `
@%s
D=M
@SP
A=M
M=D
@SP
M=M+1
`
	asm = fmt.Sprintf(asm, static)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
}

//
//  writePopSymbol
//  @Description: pop the data at the top of stack and store in segment[index]
//  @receiver cw
//  @param symbol
//  @param index
//
func (cw *CodeWriter) writePopSymbol(symbol string, index int) {
	/*
		1. pop to M register, decrease stack pointer by one.
			- @SP
			- M=M-1
		2. put 0 in D register
			- @0
			- D=A
		3. Add the stack area pointed to by LCL(local) and the D register
			- @LCL
			- D=D+M
		4. save the pointer value in R13 register
			- @R13
			- M=D
		5. put the data at the top of the stack in D register
			- @SP
			- A=M
			- D=M
		6. Add D register in the stack area pointed to by R13(temp)
			- @R13
			- A=M
			- M=D
	*/
	asm := `
@SP
M=M-1
@%d
D=A
@%s
D=D+M
@R13
M=D
@SP
A=M
D=M
@R13
A=M
M=D
`
	asm = fmt.Sprintf(asm, index, symbol)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
}

//
//  writePopRegister
//  @Description: write assemble code of pop the data at the top of stack and store in register
//  @receiver cw
//  @param index
//
func (cw *CodeWriter) writePopRegister(index int) {
	/*
		1. pop to M register, decrease stack pointer by one.
			- @SP
			- M=M-1
		2. put the data at the top of the stack in D register
			- @SP
			- A=M
			- D=M
		3. Add D register in the stack area pointed to by R11. R11 is 5 + index(=6).
			- @R11
			- M=D
	*/
	asm := `
@SP
M=M-1
@SP
A=M
D=M
@R%d
M=D
`
	asm = fmt.Sprintf(asm, index)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
}

//
//  writePopStatic
//  @Description: write assemble code of pop the data at the top of stack and store in static value
//  @receiver cw
//  @param index
//
func (cw *CodeWriter) writePopStatic(index int) {
	// set the name of static value
	static := fmt.Sprintf("%s.%d", cw.filename, index)
	/*
		1. pop to M register, decrease stack pointer by one.
			- @SP
			- M=M-1
		2. put the data at the top of the stack in D register
			- @SP
			- A=M
			- D=M
		3. Add D register in the stack area pointed to by static.
			- @StaticTest.0
			- M=D
	*/
	asm := `
@SP
M=M-1
@SP
A=M
D=M
@%s
M=D
`
	asm = fmt.Sprintf(asm, static)
	w := bufio.NewWriter(cw.writer)
	w.WriteString(asm)
	w.Flush()
}

File: 07-VM-Translator1/java/src/CodeWriter.java
import java.io.*;

/**
 * @Author HuangChuan
 * @Create in 2021/12/09 20:52
 */
public class CodeWriter {
	//the source file
	private File file;
	private File outputFile;
	BufferedWriter bufferedWriter;
	//this variable is used to distinguish the L-Command
	private int i = 0;

	public BufferedWriter getBufferedWriter() {
		return bufferedWriter;
	}

	public CodeWriter(File file) throws FileNotFoundException {
		this.file = file;
		setFileName(file.getPath());
		this.bufferedWriter = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(outputFile)));
	}

	/*
	 * @Author HuangChuan
	 * @Description //get the output file path by input file path
	 * @Date 21:35 2021/12/9
	 * @Param [fileName]
	 * @return void
	 **/
	public void setFileName(String fileName) {
		String outputFilePath = fileName.substring(0, fileName.lastIndexOf("\\")) + "\\" + fileName.substring(fileName.lastIndexOf("\\") + 1, fileName.indexOf(".")) + ".asm";
		this.outputFile = new File(outputFilePath);
	}

	/*
	 * @Author HuangChuan
	 * @Description //the assemble code of get top value of stack
	 * @Date 11:38 2021/12/11
	 * @Param []
	 * @return void
	 **/
	private void pop() throws IOException {
		this.bufferedWriter.write("//get the top value of stack\r\n"
				+ "@SP\r\n"
				+ "M=M-1\r\n"
				+ "A=M\r\n"
				+ "D=M\r\n");

	}

	/*
	 * @Author HuangChuan
	 * @Description //the assemble code of push value into stack
	 * @Date 12:42 2021/12/11
	 * @Param [value]
	 * @return void
	 **/
	private void push(String value) throws IOException {
		this.bufferedWriter.write("//push the value into stack\r\n"
				+ "@SP\r\n"
				+ "A=M\r\n"
				+ "M=" + value + "\r\n"
				+ "@SP\r\n"
				+ "M=M+1\r\n");
	}

	/*
	 * @Author HuangChuan
	 * @Description //compute the address by basic address, and get the value in the unit by address
	 * @Date 18:57 2021/12/11
	 * @Param [arg1: which segment, arg2: the offset]
	 * @return void
	 **/
	private void pushByArg(String arg1, String arg2) throws IOException {
		this.bufferedWriter.write("@" + arg1 + "\r\n"
				+ "D=M\r\n"
				+ "@" + arg2 + "\r\n"
				+ "A=A+D\r\n"
				+ "D=M\r\n");
		push("D");
	}

	/*
	 * @Author HuangChuan
	 * @Description //store the temporary value int general register
	 * @Date 13:13 2021/12/11
	 * @Param [reg]
	 * @return void
	 **/
	private void storeInReg(String reg) throws IOException {
		this.bufferedWriter.write("//store the result temporarily\r\n"
				+ "@" + reg + "\r\n"
				+ "M=D\r\n");
	}

	/*
	 * @Author HuangChuan
	 * @Description //get the two top values of stack and store them in R13 and R14
	 * @Date 13:17 2021/12/11
	 * @Param []
	 * @return void
	 **/
	private void getTwoTopElemAndStoreInReg() throws IOException {
		//get the first value
		pop();
		storeInReg("R14");
		//get the second value
		pop();
		storeInReg("R13");
	}

	/*
	 * @Author HuangChuan
	 * @Description //the R13 store the address of the value, and D store the data, so this
	 * method is to store the top value into memory
	 * @Date 13:23 2021/12/11
	 * @Param []
	 * @return void
	 **/
	private void storeRegValueInMem() throws IOException {
		this.bufferedWriter.write("//store the top value into memory\r\n"
				+ "@R13\r\n"
				+ "A=M\r\n"
				+ "M=D\r\n");
	}

	/*
	 * @Author HuangChuan
	 * @Description //get the value of R13
	 * @Date 13:43 2021/12/11
	 * @Param []
	 * @return void
	 **/
	private void getValueOfR13() throws IOException {
		this.bufferedWriter.write("@R13\r\n"
				+ "D=M\r\n");
	}

	/*
	 * @Author HuangChuan
	 * @Description //add value in R13 and value in R14
	 * @Date 14:11 2021/12/11
	 * @Param []
	 * @return void
	 **/
	private void R13AddR14() throws IOException {
		getValueOfR13();
		this.bufferedWriter.write("@R14\r\n"
				+ "D=D+M\r\n");
	}

	/*
	 * @Author HuangChuan
	 * @Description //let value in R13 minus value in R14
	 * @Date 14:19 2021/12/11
	 * @Param []
	 * @return void
	 **/
	private void R13MinusR14() throws IOException {
		getValueOfR13();
		this.bufferedWriter.write("@R14\r\n"
				+ "D=D-M\r\n");
	}

	/*
	 * @Author HuangChuan
	 * @Description //the repeating code of writeArithmetic() method
	 * @Date 16:24 2021/12/11
	 * @Param [parser, type]
	 * @return void
	 **/
	private void switchEQ(Parser parser, String type) throws IOException {
		this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
		getTwoTopElemAndStoreInReg();
		R13MinusR14();
		this.bufferedWriter.write("@" + type + i + "\r\n"
				+ "D;J" + type + "\r\n");
		push("0");
		this.bufferedWriter.write("@END" + type + i + "\r\n"
				+ "0;JMP\r\n"
				+ "(" + type + i + ")\r\n");
		//1 represents true, 0 represents false
		push("1");
		this.bufferedWriter.write("(END" + type + i + ")\r\n");
		i++;
		this.bufferedWriter.write("\r\n");
	}

	private void switchAndOrNot(Parser parser, String type) throws IOException {
		this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
		getTwoTopElemAndStoreInReg();
		getValueOfR13();
		this.bufferedWriter.write("@R14\r\n"
				+ "D=D" + type + "M\r\n");
		push("D");
		this.bufferedWriter.write("\r\n");
	}

	/*
	 * @Author HuangChuan
	 * @Description //write arithmetic command
	 * @Date 17:38 2021/12/11
	 * @Param [parser]
	 * @return void
	 **/
	public void writeArithmetic(Parser parser) throws IOException {
		//get the command type
		String type = parser.arg1();
		switch (type) {
			case "add":
				this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
				getTwoTopElemAndStoreInReg();
				R13AddR14();
				push("D");
				this.bufferedWriter.write("\r\n");
				break;
			case "sub":
				this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
				getTwoTopElemAndStoreInReg();
				R13MinusR14();
				push("D");
				this.bufferedWriter.write("\r\n");
				break;
			case "eq":
				switchEQ(parser, "EQ");
				break;
			case "gt":
				switchEQ(parser, "GT");
				break;
			case "lt":
				switchEQ(parser, "LT");
				break;
			case "and":
				switchAndOrNot(parser, "&");
				break;
			case "or":
				switchAndOrNot(parser, "|");
				break;
			case "not":
				switchAndOrNot(parser, "!");
				break;
			case "neg":
				this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
				pop();
				this.bufferedWriter.write("@0\r\n"
						+ "D=A-D\r\n");
				push("D");
				this.bufferedWriter.write("\r\n");
				break;
			default:
				break;
		}
	}

	/*
	 * @Author HuangChuan
	 * @Description //write pop and push command
	 * @Date 13:33 2021/12/12
	 * @Param [parser]
	 * @return void
	 **/
	public void writePushPop(Parser parser) throws IOException {
		CommandType commandType = parser.getCommandType();
		if (commandType.equals(CommandType.C_PUSH)) {
			String arg1 = parser.arg1();
			String arg2 = parser.arg2();
			switch (arg1) {
				case "argument":
					this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
					pushByArg("ARG", arg2);
					this.bufferedWriter.write("\r\n");
					break;
				case "local":
					this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
					pushByArg("LCL", arg2);
					this.bufferedWriter.write("\r\n");
					break;
				case "static":
					this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
					this.bufferedWriter.write("@" + this.file.getName().substring(0,
							this.file.getName().indexOf(".") + 1) + arg2 + "\r\n"
							+ "D=M\r\n");
					push("D");
					this.bufferedWriter.write("\r\n");
					break;
				case "constant":
					this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
					this.bufferedWriter.write("@" + arg2 + "\r\n"
							+ "D=A\r\n");
					push("D");
					this.bufferedWriter.write("\r\n");
					break;
				case "this":
					this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
					pushByArg("THIS", arg2);
					this.bufferedWriter.write("\r\n");
					break;
				case "that":
					this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
					pushByArg("THAT", arg2);
					this.bufferedWriter.write("\r\n");
					break;
				case "pointer":
					this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
					this.bufferedWriter.write("@THIS\r\n"
							+ "D=A\r\n"
							+ "@" + arg2 + "\r\n"
							+ "A=A+D\r\n"
							+ "D=M\r\n");
					push("D");
					this.bufferedWriter.write("\r\n");
					break;
				case "temp":
					this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
					this.bufferedWriter.write("@5\r\n"
							+ "D=A\r\n"
							+ "@" + arg2 + "\r\n"
							+ "A=A+D\r\n"
							+ "D=M\r\n");
					push("D");
					this.bufferedWriter.write("\r\n");
					break;
				default:
					break;
			}
		} else if (commandType.equals(CommandType.C_POP)) {
			String arg1 = parser.arg1();
			String arg2 = parser.arg2();
			switch (arg1) {
				case "argument":
					this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
					this.bufferedWriter.write("@ARG\r\n"
							+ "D=M\r\n"
							+ "@" + arg2 + "\r\n"
							+ "D=D+A\r\n");
					storeInReg("R13");
					pop();
					storeRegValueInMem();
					this.bufferedWriter.write("\r\n");
					break;
				case "local":
					this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
					this.bufferedWriter.write("@LCL\r\n"
							+ "@" + arg2 + "\r\n"
							+ "D=D+A\r\n");
					storeInReg("R13");
					pop();
					storeRegValueInMem();
					this.bufferedWriter.write("\r\n");
					break;
				case "static":
					this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
					pop();
					this.bufferedWriter.write("@" + this.file.getName().substring(0,
							this.file.getName().indexOf(".") + 1) + arg2 + "\r\n"
							+ "M=D\r\n");
					this.bufferedWriter.write("\r\n");
					break;
				case "constant":
					//pop constant is not exist
					break;
				case "this":
					this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
					this.bufferedWriter.write("@THIS\r\n"
							+ "D=M\r\n"
							+ "@" + arg2 + "\r\n"
							+ "D=D+A\r\n");
					storeInReg("R13");
					pop();
					storeRegValueInMem();
					this.bufferedWriter.write("\r\n");
					break;
				case "that":
					this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
					this.bufferedWriter.write("@THAT\r\n"
							+ "D=M\r\n"
							+ "@" + arg2 + "\r\n"
							+ "D=D+A\r\n");
					storeInReg("R13");
					pop();
					storeRegValueInMem();
					this.bufferedWriter.write("\r\n");
					break;
				case "pointer":
					this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
					this.bufferedWriter.write("@THIS\r\n"
							+ "D=A\r\n"
							+ "@" + arg2 + "\r\n"
							+ "D=D+A\r\n");
					storeInReg("R13");
					pop();
					storeRegValueInMem();
					this.bufferedWriter.write("\r\n");
					break;
				case "temp":
					this.bufferedWriter.write("//vm command:" + parser.getCurrentCommand() + "\r\n");
					this.bufferedWriter.write("@5\r\n"
							+ "D=A\r\n"
							+ "@" + arg2 + "\r\n"
							+ "D=D+A\r\n");
					storeInReg("R13");
					pop();
					storeRegValueInMem();
					this.bufferedWriter.write("\r\n");
					break;
				default:
					break;
			}
		} else {
			System.out.println("this command is not push or pop");
		}
	}
	/*
	 * @Author HuangChuan
	 * @Description //close the output file stream
	 * @Date 13:56 2021/12/12
	 * @Param []
	 * @return void
	 **/
	public void close() throws IOException {
		this.bufferedWriter.close();
	}
}
